tool/tctl/common/top_command.go

```diff
@@ -336,6 +391,124 @@ type Report struct {
 	Cache BackendStats
 	// Cluster is cluster stats
 	Cluster ClusterStats
+	// Watcher is watcher stats
+	Watcher *WatcherStats
+}
+
+// WatcherStats contains watcher stats
+type WatcherStats struct {
+	// EventSize is an event size histogram
+	EventSize Histogram
+	// TopEvents is a collection of resources to their events
+	TopEvents map[string]Event
+	//EventsPerSecond is the events per sec buffer
+	EventsPerSecond *CircularBuffer
+	// BytesPerSecond is the bytes per sec buffer
+	BytesPerSecond *CircularBuffer
+}
+
+// CircularBuffer implements an in-memory circular buffer of predefined size
+type CircularBuffer struct {
+	sync.Mutex
+	buf   []float64
+	start int
+	end   int
+	size  int
+}
+
+// NewCircularBuffer returns a new instance of a circular buffer that will hold
+// size elements before it rotates
+func NewCircularBuffer(size int) (*CircularBuffer, error) {
+	if size <= 0 {
+		return nil, trace.BadParameter("circular buffer size should be > 0")
+	}
+	buf := &CircularBuffer{
+		buf:   make([]float64, size),
+		start: -1,
+		end:   -1,
+		size:  0,
+	}
+	return buf, nil
+}
+
+// Data returns the data in the correct order
+func (t *CircularBuffer) Data(count int) []float64 {
+	t.Lock()
+	defer t.Unlock()
+
+	if t.size == 0 {
+		return nil
+	}
+
+	// skip first N items so that the most recent are always displayed
+	start := t.start
+	if count < t.size {
+		start = t.start + (t.size - count)
+	}
+
+	if start <= t.end {
+		return t.buf[start : t.end+1]
+	}
+
+	return append(t.buf[start:], t.buf[:start]...)
+}
```

Comment:
The current code is incorrect and can produce defective outputs or lead to "slice bounds out of range" panics. A possible fix:
```suggestion
	if count <= 0 || t.size == 0 {
		return nil
	}

	// skip first N items so that the most recent are always displayed
	start := t.start
	if count < t.size {
		start = (t.start + (t.size - count)) % len(t.buf)
	}

	if start <= t.end {
		return t.buf[start : t.end+1]
	}

	return append(t.buf[start:], t.buf[:t.end+1]...)
}
```